### PR TITLE
USCensusPEP_By_SexRaceHispanicOrigin changes done

### PR DESCRIPTION
**Pull request by kurus21**

### CODE REVIEW

File: scripts/us_census/pep/pep_by_srh/download.sh
Comment (ajaits) [inline scripts/us_census/pep/pep_by_srh/download.sh:1]: Is this file still needed? 

curl is not supported on the autorefresh framework.

File: scripts/us_census/pep/pep_by_srh/download.sh [hunk 1]
```diff
@@ -2,76 +2,89 @@
...
 for file in $(cat download_county_files_1990_2000.txt)
 do
     $(curl https://www2.census.gov/programs-surveys/popest/datasets/1990-2000/counties/asrh/${file} \
-          -o ./download_files/1990_2000/county/${file})
+          -o ./input_files/1990_2000/county/${file})
+done
+
+for file in $(cat download_county_files_2020_2023.txt)
+do
+    $(curl https://www2.census.gov/programs-surveys/popest/datasets/2020-2023/counties/asrh/${file} \
```
Comment (ajaits) [inline scripts/us_census/pep/pep_by_srh/download.sh:61]: no need to wrap command with `$(...)`
Also safer to enclose url and arguments in double quotes `""`.

can do
```
curl "https://www2.census.gov/programs-surveys/popest/datasets/2020-2023/counties/asrh/${file}" \
  "/input_files/2020_2023/county/${file}"
```

File: scripts/us_census/pep/pep_by_srh/process.py [hunk 2]
```diff
@@ -865,11 +1234,19 @@ def _consolidate_county_files():
             lambda r: _calculate_asis_measure_method(r.YEAR, r.SV), axis=1)
 
         if file == county_file[0]:
+            #added by Shamim to keep last values
```
Comment (ajaits) [inline scripts/us_census/pep/pep_by_srh/process.py:1237]: Pls explain the reason for drop_duplicates() here and below.

File: scripts/us_census/pep/pep_by_srh/download.sh
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/download.sh:1]: No Ajai, we are not using download.sh file anymore.

File: scripts/us_census/pep/pep_by_srh/download.sh (hunk 1 above)
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/download.sh:61]: Thanks for checking Ajai. We've moved away from using download.sh

File: scripts/us_census/pep/pep_by_srh/process.py (hunk 2 above)
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/process.py:1237]: Sure, will need some time to analyze this. I'll keep you posted.

File: scripts/us_census/pep/pep_by_srh/process.py [hunk 3]
```diff
@@ -43,13 +47,23 @@
...
 import os
 import pandas as pd
+import requests
+import shutil
+import json
+import time
+from datetime import datetime as dt
+from absl import logging
+
+import logging
```
Comment (krishnaswamypradeep) [inline scripts/us_census/pep/pep_by_srh/process.py:64]: Remove this default logging module.

File: scripts/us_census/pep/pep_by_srh/process.py (hunk 2 above)
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/process.py]: After drop_duplicates(), got some point deletions for eight places under 2020 year in differ(similar to Harsha's case). As I have considered the latest '2020' year data.

File: scripts/us_census/pep/pep_by_srh/process.py (hunk 3 above)
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/process.py]: removed.

File: scripts/us_census/pep/pep_by_srh/process.py
Comment (krishnaswamypradeep) [inline scripts/us_census/pep/pep_by_srh/process.py:1]: In this script, there are several instances where DataFrames are dropped using hardcoded column lists. Please refactor the code to dynamically identify and retain only the necessary columns for processing.

File: scripts/us_census/pep/pep_by_srh/process.py [hunk 4]
```diff
@@ -638,6 +995,7 @@ def _process_county_files(download_dir):
     _process_county_files_1990_2000(download_dir)
     _process_county_files_2000_2010(download_dir)
     _process_county_files_2010_2020(download_dir)
+    _process_county_files_2020_2029(download_dir)
 
 
 def _consolidate_national_files():
```
Comment (krishnaswamypradeep) [inline scripts/us_census/pep/pep_by_srh/process.py:1002]: Implement try/catch block in this method

File: scripts/us_census/pep/pep_by_srh/process.py [hunk 5]
```diff
@@ -837,16 +1200,18 @@ def _consolidate_state_files():
 
 def _consolidate_county_files():
```
Comment (krishnaswamypradeep) [inline scripts/us_census/pep/pep_by_srh/process.py:1202]: Implement try/catch block in this method.

**Review by krishnaswamypradeep — COMMENTED** [review]
Please work on the comment provided in the script file.

File: scripts/us_census/pep/pep_by_srh/process.py
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/process.py:1]: Addressed.

File: scripts/us_census/pep/pep_by_srh/process.py (hunk 5 above)
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/process.py:1352]: Implemented try/catch block

File: scripts/us_census/pep/pep_by_srh/process.py [hunk 6]
```diff
@@ -43,13 +47,19 @@
 input files, creates necessary folders for processing.
```
Comment (ajaits) [inline scripts/us_census/pep/pep_by_srh/process.py:47]: Pls update the download instructions as download.sh is deleted.

File: scripts/us_census/pep/pep_by_srh/process.py [hunk 7]
```diff
@@ -971,7 +1566,34 @@ def _consolidate_files():
...
+    global _FILES_TO_DOWNLOAD
+    with open(os.path.join(_MODULE_DIR, 'input_url.json'), 'r') as input_file:
+        _FILES_TO_DOWNLOAD = json.load(input_file)
+
+    urls_to_scan = [
+        "https://www2.census.gov/programs-surveys/popest/datasets/{YEAR}/counties/asrh/cc-est{YEAR}-alldata.csv"
+    ]
+
+    # This method will generate URLs for the years 2024 to 2029
+    for future_year in range(2024, 2030):
```
Comment (ajaits) [inline scripts/us_census/pep/pep_by_srh/process.py:1582]: Can you try with url for 2023 also generated dynamically?

File: scripts/us_census/pep/pep_by_srh/process.py [hunk 8]
```diff
@@ -971,7 +1566,34 @@ def _consolidate_files():
...
+        if dt.now().year > future_year:
+            YEAR = future_year
+            download_path = f"20{YEAR}_{YEAR+1}/county/cc-est{YEAR}-alldata.csv"  # Use f-string for dynamic path
```
Comment (ajaits) [inline scripts/us_census/pep/pep_by_srh/process.py:1585]: Should the base year be `/2020_2023/` instead of `/2022_2023/`?

File: scripts/us_census/pep/pep_by_srh/process.py [hunk 9]
```diff
@@ -998,27 +1620,116 @@ def _create_output_n_process_folders():
...
+    session = requests.session()
+    max_retry = 5
+
+    for file_to_download in _FILES_TO_DOWNLOAD:
+        file_name_to_save = None
+        url = file_to_download['download_path']
+
+        # Determine the filename to save the downloaded file
+        if 'file_name' in file_to_download and len(
+                file_to_download['file_name']) > 5:
```
Comment (ajaits) [inline scripts/us_census/pep/pep_by_srh/process.py:1670]: Pls explain the check for file name length of 5. Why are smaller file names and issue?

Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/process.py:1670]: Addressed.

File: scripts/us_census/pep/pep_by_srh/process.py (hunk 8 above)
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/process.py:1585]: Updated.

File: scripts/us_census/pep/pep_by_srh/process.py (hunk 7 above)
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/process.py:1582]: Yes, updated to 2023 and got the output.

File: scripts/us_census/pep/pep_by_srh/process.py (hunk 6 above)
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/process.py:47]: Updated.

File: scripts/us_census/pep/pep_by_srh/process.py [hunk 10]
```diff
@@ -985,8 +1607,8 @@ def _process_files(download_dir):
     """
     _process_county_files(download_dir)
     _process_state_files(download_dir)
-    # _process_geo_level_aggregations will process state 2000 - 2020 data
-    # and national 1980 - 2020 data
+    # _process_geo_level_aggregations will process state 2000 - 2020, 2020 - 2029 data
+    # and national 1980 - 2020, 2020 - 2029 data
```
[the remark this answers is not in the document]
Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/process.py:1440]: Hi Ajai,
Although national files exist, they may require additional processing compared to state and county data. If needed, we can include the necessary script changes to handle the processing of national files.
We have tried to keep the script structure as close to the original as possible, with minimal changes.

File: scripts/us_census/pep/pep_by_srh/input_url.json [hunk 11]
```diff
@@ -0,0 +1,262 @@
+[
+    {
+        "download_path": "https://www2.census.gov/programs-surveys/popest/datasets/1990-2000/counties/asrh/co-99-10.txt",
+        "file_path": "1990_2000/county/"
+    },
+    {
+        "download_path": "https://www2.census.gov/programs-surveys/popest/datasets/2020-2023/counties/asrh/cc-est2023-alldata.csv",
```
Comment (krishnaswamypradeep) [inline scripts/us_census/pep/pep_by_srh/input_url.json:7]: Remove this link as it should be dynamically constructed using the future year URL.

Comment (Girish3320) [reply scripts/us_census/pep/pep_by_srh/input_url.json]: Done.

**Review by krishnaswamypradeep — APPROVED** [review]
Thanks Girish. Looks good.

**Review by ajaits — APPROVED** [review]
